src/libsyntax/diagnostics/plugin.rs

```diff
@@ -132,17 +117,17 @@ pub fn expand_register_diagnostic<'cx>(ecx: &'cx mut ExtCtxt,
         }
     });
     // Add the error to the map.
-    with_registered_diagnostics(|diagnostics| {
-        let info = ErrorInfo {
-            description,
-            use_site: None
-        };
-        if diagnostics.insert(code.name, info).is_some() {
-            ecx.span_err(span, &format!(
-                "diagnostic code {} already registered", code
-            ));
-        }
-    });
+    let mut diagnostics = ecx.parse_sess.registered_diagnostics.lock();
```

Comment:
Could we add a `with` method to `Lock`? That would make it clearer in many cases whether the lock is held until the end of the block intentionally or not.

Reply:
That's probably not a bad idea. `with` sounds a little innocent for a lock though.

Reply:
`with_lock()` maybe?